src/status/pollingClient.ts

```diff
@@ -60,74 +54,34 @@ export class PollingClient extends AsyncOptionalCreatable<PollingClient.Options>
    * Returns a promise to call the specified polling function using the interval and timeout specified
    * in the polling options.
    */
-  public subscribe(): Promise<AnyJson | undefined> {
-    // This promise is held open while setInterval tries to resolve or reject.
-    // If set interval can't do it then the timeout will reject.
-    return new Promise((resolve, reject) => {
-      // Use set interval to periodically call the polling function
-
-      // This try catch enables support for time{0} since setInterval only supports
-      // time {1}. In other words, we should call first then wait for the first interval.
-      this.doPoll()
-        .then((result: Optional<StatusResult>) => {
-          if (result && result.completed) {
-            resolve(result.payload);
-          } else {
-            this.interval = setInterval(
-              // eslint-disable-next-line @typescript-eslint/no-misused-promises
-              PollingClient.prototype.doPoll.bind(this, resolve, reject),
-              this.options.frequency.milliseconds
-            );
-          }
-        })
-        .catch(() => {
-          this.interval = setInterval(
-            // eslint-disable-next-line @typescript-eslint/no-misused-promises
-            PollingClient.prototype.doPoll.bind(this, resolve, reject),
-            this.options.frequency.milliseconds
-          );
-        });
-
-      // hard polling timeout.
-      this.timeout = setTimeout(() => {
-        this.logger.debug('Polling timed out');
-        this.clearAll();
-        // @todo should go in messages.
-        reject(new SfdxError('The client has timed out.', this.options.timeoutErrorName || 'PollingClientTimeout'));
-      }, this.options.timeout.milliseconds);
+  public async subscribe(): Promise<AnyJson | undefined> {
+    let errorInPollingFunction; // keep this around for returning in the catch block
+    const doPoll = async () => {
+      let result;
+      try {
+        result = await this.options.poll();
+      } catch (error) {
+        errorInPollingFunction = error;
+        // there was an actual error thrown, so we don't want to keep retrying
+        throw new NotRetryableError(error.name);
+      }
+      if (result.completed) {
+        return result.payload;
+      }
+      throw new Error('Operation did not complete.  Retrying...'); // triggers a retry
```

Comment:
Are there any tests for this change?

Reply:
the tests are the same (it gets hit by its own tests plus myDomainResolver).

We **could** try to simulate the SDR bug (have a slow-returning test function) to prove that calls are only made after responses come back.  Seem like a good use of time?

Reply:
it could be worth it, just to prove that we've actually fixed it, and ensure it won't happen again. Maybe timebox it?

it would take the PR from good to great

Reply:
add one UT.  Could be useful when the next person feels like swapping out the retry library, etc.